sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+        self.fetch_imds(IMDS_PINNED_DATE, "user-data", "user-data")
+    }
+
+    /// Fetch data from IMDS, preferring an override file if present.
+    pub fn fetch_imds(
+        &mut self,
+        schema_version: &str,
+        target: &str,
+        description: &str,
+    ) -> Result<Option<String>> {
+        let uri = format!("{}/{}/{}", IMDS_BASE_URI, schema_version, target);
+        debug!("Requesting {} from {}", description, &uri);
+        let mut attempt: u16 = 0;
+        loop {
+            attempt += 1;
+            ensure!(attempt <= 2, error::FailedFetch { attempt });
+            let response = self
+                .client
+                .get(&uri)
+                .header("X-aws-ec2-metadata-token", &self.session_token)
+                .send()
+                .context(error::Request {
+                    method: "GET",
+                    uri: &uri,
+                })?;
+            trace!("IMDS response: {:?}", &response);
+
+            match response.status() {
+                code @ StatusCode::OK => {
+                    info!("Received {}", description);
+                    let response_body = response.text().context(error::ResponseBody {
+                        method: "GET",
+                        uri: &uri,
+                        code,
+                    })?;
+                    trace!("Response text: {:?}", &response_body);
+
+                    return Ok(Some(String::from(response_body.trim_end())));
+                }
+
+                // IMDS returns 404 if no user data is given, or if IMDS is disabled
+                StatusCode::NOT_FOUND => return Ok(None),
+
+                // IMDS returns 401 if the session token is expired or invalid
+                StatusCode::UNAUTHORIZED => {
+                    warn!("Session token is invalid or expired");
+                    self.refresh_token()?;
+                    info!("Refreshed session token");
+                    continue;
+                }
+
+                code => {
+                    let response_body = response.text().context(error::ResponseBody {
+                        method: "GET",
+                        uri: &uri,
+                        code,
+                    })?;
+                    trace!("Response text: {:?}", &response_body);
+
+                    return error::Response {
+                        method: "GET",
+                        uri: &uri,
+                        code,
+                        response_body,
+                    }
+                    .fail();
+                }
+            }
+        }
+    }
+
+    /// Fetches a new session token and adds it to the current ImdsClient.
+    pub fn refresh_token(&mut self) -> Result<()> {
+        self.session_token = fetch_token(&self.client)?;
+        Ok(())
+    }
+}
+
+/// Helper to fetch an IMDSv2 session token that is valid for 60 seconds.
+pub fn fetch_token(client: &Client) -> Result<String> {
```

Comment:
Is fetch_token actually part of the public interface for this library? I don't see it being used in the clients. If it's not used in the clients maybe we should make it private.

Reply:
In my mind, ideally, the user doesn't know or care if they're using tokens or IMDSv2 or IMDSv1, unless they ask...

Reply:
It was in the first iteration, but now it's no longer necessary. I'll remove it.